cvat/apps/events/handlers.py

```diff
@@ -594,3 +588,99 @@ def handle_viewset_exception(exc, context):
     )
 
     return response
+
+def handle_client_events_push(request, data: dict):
+    TIME_THRESHOLD = datetime.timedelta(seconds=100)
+    WORKING_TIME_SCOPE = 'send:working_time'
+    WORKING_TIME_RESOLUTION = datetime.timedelta(milliseconds=1)
+    COLLAPSED_EVENT_SCOPES = frozenset(("change:frame",))
+    org = request.iam_context["organization"]
+
+    def read_ids(event: dict) -> tuple[int | None, int | None, int | None]:
+        return event.get("job_id"), event.get("task_id"), event.get("project_id")
+
+    def get_end_timestamp(event: dict) -> datetime.datetime:
+        if event["scope"] in COLLAPSED_EVENT_SCOPES:
+            return event["timestamp"] + datetime.timedelta(milliseconds=event["duration"])
+        return event["timestamp"]
+
+    def generate_wt_event(
+        ids: tuple[int | None, int | None, int | None],
+        wt: datetime.timedelta,
+        common: dict
+    ) -> EventSerializer | None:
+        job_id, task_id, project_id = ids
+        if wt.total_seconds():
+            value = wt // WORKING_TIME_RESOLUTION
+            event = EventSerializer(data={
+                **common,
+                "scope": WORKING_TIME_SCOPE,
+                "obj_name": "working_time",
+                "obj_val": value,
+                "source": "server",
+                "count": 1,
+                "project_id": project_id,
+                "task_id": task_id,
+                "job_id": job_id,
+                # keep it in payload for backward compatibility
+                # but in the future it is much better to use a dedicated field "obj_val"
+                # because parsing JSON in SQL query is very slow
+                "payload": json.dumps({ "working_time": value })
+            })
+
+            return event
+
+    if previous_event := data["previous_event"]:
+        previous_end_timestamp = get_end_timestamp(previous_event)
+        previous_ids = read_ids(previous_event)
+    elif data["events"]:
+        previous_end_timestamp = data["events"][0]["timestamp"]
+        previous_ids = read_ids(data["events"][0])
+
+    working_time_per_ids = {}
+    for event in data["events"]:
+        working_time = datetime.timedelta()
+        timestamp = event["timestamp"]
+
+        if timestamp > previous_end_timestamp:
+            t_diff = timestamp - previous_end_timestamp
+            if t_diff < TIME_THRESHOLD:
+                working_time += t_diff
+
+            previous_end_timestamp = timestamp
+
+        end_timestamp = get_end_timestamp(event)
+        if end_timestamp > previous_end_timestamp:
+            working_time += end_timestamp - previous_end_timestamp
+            previous_end_timestamp = end_timestamp
+
+        if previous_ids not in working_time_per_ids:
+            working_time_per_ids[previous_ids] = {
+                "value": datetime.timedelta(),
+                "timestamp": timestamp,
```

Comment:
I think it would be more accurate to initialize `timestamp` to `datetime()` here and for each event do:

```python
if previous_end_timestamp > working_time_per_ids[previous_ids]["timestamp"]:
    working_time_per_ids[previous_ids]["timestamp"] = previous_end_timestamp
```

That way, the timestamp of the working time event is the end timestamp of the last event that was used to determine it - which is the earliest time at which the working time would be known.

Reply:
I do not think we should use the last event.
There is a problem with timestamp correction (request time added to the event timestamp).

As discussed with @azhavoro , we need this correction, because it allows to sync time between client and server.
On the other way if request took long (e.g. 30 seconds), we will get the same problem with event timestamp >= Job.updated_date. It looks like a problem.

But using the first event in the sequence allows to minimize effect of this bug

Reply:
Okay. Also, I see that you changed the contents of the event to record the working time in the duration field. I like that, now the timestamp of the first original event makes sense as the timestamp of the working time event, since we can now view the WT event as spanning from the start of the first original event to the end of the last one.